kernel/exsubst.ml

```diff
@@ -0,0 +1,119 @@
+open Basic
+open Format
+open Term
+
+(** An extended substitution is a function mapping
+    - a variable (location, identifier and DB index)
+    - applied to a given number of arguments
+    - under a given number of lambda abstractions
+    to a term.
+    A substitution raises Not_found meaning that the variable is not subsituted. *)
+type ex_substitution = loc -> ident -> int -> int -> int -> term
+
+(** [apply_exsubst subst n t] applies [subst] to [t] under [n] lambda abstractions.
+      - Variables with DB index [k] <  [n] are considered "locally bound" and are never substituted.
+      - Variables with DB index [k] >= [n] may be substituted if [k-n] is mapped in [sigma]
+          and if they occur applied to enough arguments (substitution's arity). *)
+let apply_exsubst (subst:ex_substitution) (n:int) (te:term) : term*bool =
```

Comment:
Why this function cannot be implemented by one of the subst function? Why this function is not in the module Subst?

Reply:
Ok, I rephrase, why do we have two modules for substitution now? Why don't we keep only one module for substitutions?